libbpf-tools/doublefree.bpf.c

```diff
@@ -0,0 +1,165 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/* Copyright (c) 2022 LG Electronics */
+#include <vmlinux.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include "doublefree.h"
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, struct doublefree_info_t);
+} allocs SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, u64);
+} memptrs SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__uint(max_entries, MAX_ENTRIES);
+	__uint(key_size, sizeof(u32));
+} stack_traces SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, u32);
+} first_deallocs SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10);
+	__type(key, u64);
+	__type(value, u32);
+} second_deallocs SEC(".maps");
+
+int gen_alloc_exit2(struct pt_regs *ctx, u64 address)
+{
+	struct doublefree_info_t info = {0, };
+
+	if (address != 0) {
+		info.stack_id = bpf_get_stackid(ctx, &stack_traces,
+			BPF_F_USER_STACK);
+		info.is_available = 1;
+		bpf_map_update_elem(&allocs, &address, &info, BPF_ANY);
+	}
+	return 0;
+}
+
+int gen_alloc_exit(struct pt_regs *ctx)
+{
+	return gen_alloc_exit2(ctx, PT_REGS_RC(ctx));
+}
+
+int gen_free_enter(struct pt_regs *ctx, void *address)
+{
+	int stack_id = 0;
+	u64 addr = (u64)address;
+	struct doublefree_info_t *info = bpf_map_lookup_elem(&allocs, &addr);
+
+	if (!info)
+		return 0;
+
+	info->is_available -= 1;
+	stack_id = bpf_get_stackid(ctx, &stack_traces, BPF_F_USER_STACK);
+	if (info->is_available == 0) {
+		bpf_map_update_elem(&first_deallocs, &addr, &stack_id, BPF_ANY);
+	} else if (info->is_available < 0) {
+		bpf_map_update_elem(&second_deallocs, &addr, &stack_id, BPF_ANY);
+	} else {
+		bpf_printk("This code should not be printed at any rate\n");
+		bpf_printk("Please check if something goes wrong\n");
```

Comment:
I didn't know that `bpf_printk` could cause `dmesg` warning. Thanks for the information.

I agree with your opinion and I hope to get other's opinion as well before coding.
Fortunately(?) I found the similar code in existing `memleak` code.
So let's discuss this issue in already merged existing code.
I'll make a new issue and mention this comment and you.

Reply:
On x86_64 machine Ubuntu 22.04.3 LTS,
I ran `doublefree` which contains `bpf_printk` but I cannot get warning message on `dmesg`.
And I ran `memleak` which contains `bpf_printk` as well(as I mentioned previously), but I still cannot get warning message on `dmesg`.
Could you please tell me the way to get warning message on `dmesg` when `bpf_printk` compiled eBPF code is loading?

Reply:
Thank you for the checking!
`bpf_printk` is defined in file `bpftool/libbpf/src/bpf_helpers.h` and calls `bpf_trace_printk`.
So I guess we could add `bpf_printk_err` or `bpf_printk_dbg` to the file `bpf_helpers.h` and make it compiled only when `ERROR` or `DEBUG` macro is defined.
Since the file belongs to [libbpf](https://github.com/libbpf/libbpf), I'll make issue on this project.
Or you can make an issue(it's your idea).

Reply:
I changed this program to use `bpf_perf_event`. And I carry error code on event instead of print it inside bpf code. So this issue is resolved.